common/namespace/registry.go

```diff
@@ -416,54 +423,87 @@ UpdateLoop:
 
 	// NOTE: READ REF BEFORE MODIFICATION
 	// ref: historyEngine.go registerNamespaceFailoverCallback function
-	r.callbackLock.Lock()
-	defer r.callbackLock.Unlock()
-	r.triggerNamespaceChangePrepareCallbackLocked()
-	r.cacheByID.Store(newCacheByID)
-	r.cacheNameToID.Store(newCacheNameToID)
-	r.triggerNamespaceChangeCallbackLocked(prevEntries, nextEntries)
-
-	// only update last refresh time when refresh succeeded
-	r.lastRefreshTime.Store(now)
+	r.makeVisibleWithCallbacks(func() (Namespaces, Namespaces) {
+		r.cacheLock.Lock()
+		defer r.cacheLock.Unlock()
+		r.cacheByID = newCacheByID
+		r.cacheNameToID = newCacheNameToID
+		return prevEntries, nextEntries
+	})
 	return nil
 }
 
-func (r *registry) updateNameToIDCache(
-	cacheNameToID cache.Cache,
-	name string,
-	id string,
-) {
-	cacheNameToID.Put(name, id)
+func (r *registry) updateNameToIDCache(c cache.Cache, name string, id string) {
+	c.Put(name, id)
 }
 
 func (r *registry) updateIDToNamespaceCache(
 	cacheByID cache.Cache,
 	id string,
-	record *Namespace,
+	newNS *Namespace,
 ) *Namespace {
-	prevCacheRec := cacheByID.Put(id, record)
-	if e, ok := prevCacheRec.(*Namespace); ok &&
-		record.notificationVersion > e.notificationVersion &&
+	prevCacheRec := cacheByID.Put(id, newNS)
+	if existingNS, ok := prevCacheRec.(*Namespace); ok &&
+		newNS.notificationVersion > existingNS.notificationVersion &&
 		r.globalNamespacesEnabled {
-		return e
+		return existingNS
 	}
 	return nil
 }
 
 // getNamespace retrieves the information from the cache if it exists
 func (r *registry) getNamespace(name string) (*Namespace, error) {
-	if id, ok := r.cacheNameToID.Load().(cache.Cache).Get(name).(string); ok {
+	r.cacheLock.RLock()
+	if id, ok := r.cacheNameToID.Get(name).(string); ok {
+		r.cacheLock.RUnlock()
```

Comment:
It's kind of unfortunate that getting by name has to lock/unlock/lock/unlock on a cache hit. Do you think there's a relatively clean way to avoid it?

Reply:
Good point. I can think of a couple of ways but all a bit awkward. I should be able to come up with something

Reply:
Dug deeper - every Get on our LRU implementation is an xlock (because an LRU read is necessarily also a write and we don't use a lockfree LL) so being crafty here to merge the two lock acquisition into one isn't going to buy us anything.